msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ConfidentialClientApplication.java

```diff
@@ -93,23 +98,58 @@ private void initClientAuthentication(IClientCredential clientCredential) {
         }
     }
 
+    /**
+     * Generates a JWT-formatted assertion string based on the provided client credential. Returns null in cases where
+     * the request for that credential type would not use a JWT assertion (e.g. client secret).
+     *
+     * @param clientCredential  The client credential to use for token acquisition.
+     * @return JWT-formatted assertion string
+     */
     String getAssertionString(IClientCredential clientCredential) {
         if (clientCredential instanceof ClientCertificate) {
-            boolean useSha1 = Authority.detectAuthorityType(this.authenticationAuthority.canonicalAuthorityUrl()) == AuthorityType.ADFS;
-
-            return JwtHelper.buildJwt(
-                    clientId(),
-                    clientCertificate,
-                    this.authenticationAuthority.selfSignedJwtAudience(),
-                    sendX5c,
-                    useSha1).assertion();
+            // Check if the current assertion is null or has expired, and if so create a new one
+            if (this.assertion == null || hasJwtExpired(this.assertion)) {
+                boolean useSha1 = Authority.detectAuthorityType(this.authenticationAuthority.canonicalAuthorityUrl()) == AuthorityType.ADFS;
+
+                this.assertion = JwtHelper.buildJwt(
+                        clientId(),
+                        clientCertificate,
+                        this.authenticationAuthority.selfSignedJwtAudience(),
+                        sendX5c,
+                        useSha1).assertion();
+            }
+            return this.assertion;
         } else if (clientCredential instanceof ClientAssertion) {
             return ((ClientAssertion) clientCredential).assertion();
+        } else if (clientCredential instanceof ClientSecret) {
+            return null;
         } else {
             throw new IllegalArgumentException("Unsupported client credential");
         }
     }
 
+    //Overload for the common case where the application's default credential was not overridden in the request.
+    String getAssertionString() {
+        return this.getAssertionString(this.clientCredential);
+    }
+
+    /**
+     * Checks if the JWT-formatted assertion has expired by parsing the "exp" claim.
+     *
+     * @param jwt JWT string
+     * @return true if the JWT has expired. Otherwise false
+     */
+    boolean hasJwtExpired(String jwt) {
```

Comment:
A Wise man once advised me to never write optimizations before:
1.  writing tests that proved that there was something that needed to be optimized. 
2.  writing tests that actually proved the optimization where fast enough to justify the extra code.
anyway, that's my 2 pennies.